optuna/structs.py

```diff
@@ -267,6 +258,79 @@ class StudySummary(
             Datetime where the :class:`~optuna.study.Study` started.
     """
 
+    def __init__(
+            self,
+            study_name,  # type: str
+            direction,  # type: StudyDirection
+            best_trial,  # type: Optional[FrozenTrial]
+            user_attrs,  # type: Dict[str, Any]
+            system_attrs,  # type: Dict[str, Any]
+            n_trials,  # type: int
+            datetime_start,  # type: Optional[datetime]
+            study_id,  # type: int
+    ):
+        # type: (...) -> None
+
+        self.study_name = study_name
+        self.direction = direction
+        self.best_trial = best_trial
+        self.user_attrs = user_attrs
+        self.system_attrs = system_attrs
+        self.n_trials = n_trials
+        self.datetime_start = datetime_start
+        self._study_id = study_id
+
+    def __eq__(self, other):
+        # type: (Any) -> bool
+
+        if not isinstance(other, StudySummary):
+            return NotImplemented
+
+        return other.__dict__ == self.__dict__
+
+    def __ne__(self, other):
+        # type: (Any) -> bool
+
```

Comment:
We should check the type with an early return in `__ne__` too. C.f. https://github.com/optuna/optuna/pull/726#discussion_r349994303.

Reply:
Thank you for pointing it out. I fixed it and added the corresponding test case to ensure that the problem is actually fixed.

Fixed:  https://github.com/optuna/optuna/pull/718/commits/6ed2ed53fd91d067dc898e6e49eb435acc9150a2